File: scripts/dev_tools/pr_context/render_pr_helpers.py
"""PR context rendering utility helpers extracted from render module."""

from __future__ import annotations

import re
from pathlib import Path
from typing import TYPE_CHECKING

from .models import (
    CONVENTIONAL_TYPES,
    IssueDetails,
    PullRequestDetails,
    format_list,
    section,
    truncate,
)

if TYPE_CHECKING:
    from collections.abc import Iterable

    from .git import GitClient


def select_default_base(git: GitClient) -> str | None:
    """Select the first existing base ref from standard default candidates."""
    candidates = [
        "origin/main",
        "origin/master",
        "main",
        "master",
        "origin/develop",
        "develop",
    ]
    for ref in candidates:
        result = git.run(["rev-parse", "--verify", "--quiet", ref], allow_error=True)
        if result.code == 0 and result.stdout.strip():
            return ref
    return None


def format_diff_path(path_text: str | None) -> str:
    """Normalize git diff path output, including rename syntax variants."""
    if path_text is None:
        return ""
    if path_text.strip() == "":
        return path_text

    trimmed = path_text.strip().strip('"')
    trimmed = re.sub(r"\{[^{}]*\s=>\s([^{}]*)\}", r"\1", trimmed)

    arrow_match = re.match(r"^\s*(.+?)\s=>\s(.+?)\s*$", trimmed)
    if arrow_match:
        return arrow_match.group(2)
    return trimmed


def convert_numstat(numstat_text: str) -> tuple[int, int, list[str]]:
    """Convert git numstat output into totals and raw path list."""
    adds = 0
    dels = 0
    files: list[str] = []

    for raw_line in numstat_text.splitlines():
        if not raw_line.strip():
            continue

        parts = raw_line.split("\t")
        if len(parts) < 3:
            continue

        add_part, del_part, file_part = parts[0], parts[1], parts[2]
        if add_part.isdigit():
            adds += int(add_part)
        if del_part.isdigit():
            dels += int(del_part)
        files.append(file_part)

    return adds, dels, files


def extension_summary(files: Iterable[str]) -> str:
    """Summarize changed files by extension."""
    counts: dict[str, int] = {}
    for raw in files:
        name = format_diff_path(raw)
        ext = "(unknown)"
        try:
            suffix = Path(name).suffix
            ext = suffix if suffix else "(noext)"
        except ValueError:
            fallback = re.search(r"\.([A-Za-z0-9_]+)$", name)
            ext = f".{fallback.group(1)}" if fallback else "(unknown)"

        counts[ext] = counts.get(ext, 0) + 1

    lines = [f"{counts[k]:8d}  {k}" for k in sorted(counts)]
    return "\n".join(lines)


def extract_issue_references(text: str) -> list[str]:
    """Extract issue tokens like #123 and ABC-123 in encounter order."""
    if not text:
        return []
    matches = re.findall(r"(?<!\w)#\d+|\b[A-Z][A-Z0-9]+-\d+\b", text)
    seen: set[str] = set()
    ordered: list[str] = []
    for item in matches:
        if item not in seen:
            seen.add(item)
            ordered.append(item)
    return ordered


def extract_merge_pr_numbers(subjects: Iterable[str]) -> list[str]:
    """Extract merged PR numbers from commit subject lines."""
    numbers: set[str] = set()
    pattern = re.compile(r"Merge pull request #(\d+)", re.IGNORECASE)
    for subj in subjects:
        match = pattern.search(subj)
        if match:
            numbers.add(f"#{match.group(1)}")
    return sorted(numbers)


def summarize_conventional_commits(subjects: str) -> str:
    """Count conventional commit types in subject lines."""
    counts = {key: 0 for key in CONVENTIONAL_TYPES}
    counts["other"] = 0

    for line in subjects.splitlines():
        line = line.strip()
        if not line:
            continue
        match = re.match(
            r"(feat|fix|refactor|perf|docs|test|chore|build|ci|style)(\(|!|:)",
            line,
        )
        label = match.group(1) if match else "other"
        counts[label] += 1

    non_zero = [(k, v) for k, v in counts.items() if v > 0]
    if not non_zero:
        return "(no recognizable conventional commit types)"
    return "\n".join(f"{name:<9} : {value}" for name, value in non_zero)


def format_issue_details(issue: IssueDetails) -> str:
    """Render structured issue details for PR context appendix output."""
    comments_text = format_list(issue.comments, "(no comments)")
    lines = [
        section(f"Issue {issue.number}: {issue.title}"),
        f"State: {issue.state}",
        f"Author: {issue.author}",
        f"Labels: {', '.join(issue.labels) if issue.labels else '(none)'}",
        f"Assignees: {', '.join(issue.assignees) if issue.assignees else '(none)'}",
        f"Created: {issue.created_at}",
        f"Updated: {issue.updated_at}",
        "",
        truncate(issue.body, 1200),
        "",
        "Comments:",
        comments_text,
    ]
    if issue.user_story_content:
        lines.extend(
            [
                "",
                f"User story ({issue.user_story_path or 'user-story.md'}):",
                truncate(issue.user_story_content, 1200),
            ]
        )
    return "\n".join(lines)


def format_pr_details(pr: PullRequestDetails) -> str:
    """Render structured pull request details for PR context appendix output."""
    return "\n".join(
        [
            section(f"Pull Request {pr.number}: {pr.title}"),
            f"State: {pr.state}",
            f"Author: {pr.author}",
            f"Base: {pr.base_ref}",
            f"Head: {pr.head_ref}",
            f"Created: {pr.created_at}",
            f"Updated: {pr.updated_at}",
            f"Merged: {pr.merged_at or '(not merged)'}",
            f"Labels: {', '.join(pr.labels) if pr.labels else '(none)'}",
            f"Assignees: {', '.join(pr.assignees) if pr.assignees else '(none)'}",
            truncate(pr.body, 1200),
            "",
            "Auto-close issues (from this PR):",
            format_list(pr.closing_issues, "(none)"),
            "",
            "Files (first 15):",
            format_list(pr.files_changed[:15], "(none)"),
        ]
    )


def build_close_candidates_section(
    *,
    verified: list[str],
    author_asserted: list[str],
    referenced: list[str],
    verified_reason: str,
    author_reason: str,
) -> str:
    """Render close-candidate section grouped by verification source."""
    all_auto_close = set(verified + author_asserted + referenced)
    author_auto_close = sorted(all_auto_close)
    referenced_only = sorted(set(referenced) - all_auto_close)

    return "\n".join(
        [
            section("Close candidates"),
            "Auto-close issues (verified from GitHub PR metadata):",
            format_list(verified, verified_reason),
            "",
            "Auto-close issues (author asserted):",
            format_list(author_auto_close, author_reason),
            "",
            "Referenced issues (detected):",
            format_list(referenced_only, "(none)"),
        ]
    )


def extract_changed_paths(context_text: str) -> list[str]:
    """Extract changed file paths from the 'Changed files' section text."""
    paths: list[str] = []
    capture = False
    for line in context_text.splitlines():
        if line.startswith("===== Changed files"):
            capture = True
            continue
        if capture:
            if line.startswith("====="):
                break
            if line.strip() and "\t" in line:
                path_part = line.split("\t")[-1]
                paths.append(format_diff_path(path_part.strip()))
            elif line.strip():
                paths.append(format_diff_path(line.strip()))
    return paths

File: extensions/drm-copilot/src/lib/pr-context/render-pr-helpers.ts
/**
 * PR context rendering utility helpers.
 *
 * Purpose:
 *     Port of `dev_tools/pr_context/render_pr_helpers.py`. Provide the pure
 *     rendering and extraction helpers used to build the PR comparison and
 *     summary sections: base selection, diff-path normalization, numstat
 *     conversion, extension summary, issue/PR reference extraction, conventional
 *     commit summary, issue/PR detail rendering, close-candidate and autoclose
 *     sections, and changed-path extraction.
 *
 * Responsibilities:
 *     - Replicate every regex, format string, and ordering exactly so the
 *       rendered output matches the Python source.
 *     - `selectDefaultBase` accepts an injected {@link GitClient}.
 */

import {
  CONVENTIONAL_TYPES,
  type IssueDetails,
  type PullRequestDetails,
  formatList,
  section,
  truncate,
} from "./models";
import { type GitClient } from "./git-client";

/** Default base ref candidates, probed in fixed order. */
const DEFAULT_BASE_CANDIDATES = [
  "origin/main",
  "origin/master",
  "main",
  "master",
  "origin/develop",
  "develop",
] as const;

/**
 * Select the first existing base ref from standard default candidates.
 *
 * Mirrors Python `select_default_base`: probe each candidate with
 * `rev-parse --verify --quiet`, returning the first whose exit is `0` and whose
 * stdout is non-empty.
 *
 * @param git Injected git client.
 * @returns The first existing candidate ref, or `null` when none verify.
 */
export function selectDefaultBase(git: GitClient): string | null {
  // Probe each candidate ref in priority order; the first that verifies wins.
  for (const ref of DEFAULT_BASE_CANDIDATES) {
    const result = git.run(["rev-parse", "--verify", "--quiet", ref], {
      allowError: true,
    });
    if (result.code === 0 && result.stdout.trim()) {
      return ref;
    }
  }
  return null;
}

/**
 * Normalize a git diff path, including rename syntax variants.
 *
 * Mirrors Python `format_diff_path`: `null` -> `""`; an all-whitespace string is
 * returned unchanged; otherwise strip quotes, collapse `{old => new}` brace
 * renames to the new segment, and resolve `old => new` arrow renames to the new
 * path.
 *
 * @param pathText Raw diff path text.
 * @returns The normalized path.
 */
export function formatDiffPath(pathText: string | null): string {
  if (pathText === null) {
    return "";
  }
  if (pathText.trim() === "") {
    return pathText;
  }

  let trimmed = pathText.trim().replace(/^"+|"+$/gu, "");
  trimmed = trimmed.replace(/\{[^{}]*\s=>\s([^{}]*)\}/gu, "$1");

  const arrowMatch = /^\s*(.+?)\s=>\s(.+?)\s*$/u.exec(trimmed);
  if (arrowMatch) {
    return arrowMatch[2]!;
  }
  return trimmed;
}

/**
 * Convert git numstat output into totals and a raw path list.
 *
 * Mirrors Python `convert_numstat`: sum integer add/del columns, collect the
 * third tab-separated field as the path, skipping blank/short lines.
 *
 * @param numstatText Raw `git diff --numstat` output.
 * @returns A tuple of `[additions, deletions, files]`.
 */
export function convertNumstat(
  numstatText: string,
): [number, number, string[]] {
  let adds = 0;
  let dels = 0;
  const files: string[] = [];

  // Parse each non-blank numstat row into add/del totals and the file path.
  for (const rawLine of splitLines(numstatText)) {
    if (!rawLine.trim()) {
      continue;
    }
    const parts = rawLine.split("\t");
    if (parts.length < 3) {
      continue;
    }
    const [addPart, delPart, filePart] = [parts[0]!, parts[1]!, parts[2]!];
    if (isDigits(addPart)) {
      adds += Number.parseInt(addPart, 10);
    }
    if (isDigits(delPart)) {
      dels += Number.parseInt(delPart, 10);
    }
    files.push(filePart);
  }

  return [adds, dels, files];
}

/**
 * Summarize changed files by extension.
 *
 * Mirrors Python `extension_summary`: normalize each path, take its suffix
 * (`(noext)` when none), count per extension, and render `%8d  ext` lines sorted
 * by extension.
 *
 * @param files Iterable of raw file paths.
 * @returns The formatted per-extension summary.
 */
export function extensionSummary(files: Iterable<string>): string {
  const counts = new Map<string, number>();
  // Count files per extension, normalizing the diff path first.
  for (const raw of files) {
    const name = formatDiffPath(raw);
    const suffix = pathSuffix(name);
    const ext = suffix ? suffix : "(noext)";
    counts.set(ext, (counts.get(ext) ?? 0) + 1);
  }

  const lines = [...counts.keys()]
    .sort(compareCodePoint)
    .map((key) => `${padCount(counts.get(key)!)}  ${key}`);
  return lines.join("\n");
}

/**
 * Extract issue tokens like `#123` and `ABC-123` in encounter order.
 *
 * Mirrors Python `extract_issue_references`.
 *
 * @param text Source text.
 * @returns Ordered, deduplicated reference tokens.
 */
export function extractIssueReferences(text: string): string[] {
  if (!text) {
    return [];
  }
  const matches = text.match(/(?<!\w)#\d+|\b[A-Z][A-Z0-9]+-\d+\b/gu) ?? [];
  const seen = new Set<string>();
  const ordered: string[] = [];
  for (const item of matches) {
    if (!seen.has(item)) {
      seen.add(item);
      ordered.push(item);
    }
  }
  return ordered;
}

/**
 * Extract merged PR numbers from commit subject lines.
 *
 * Mirrors Python `extract_merge_pr_numbers`: match `Merge pull request #N`
 * (case-insensitive) and return sorted `#N` strings.
 *
 * @param subjects Iterable of commit subject lines.
 * @returns Sorted `#N` PR numbers.
 */
export function extractMergePrNumbers(subjects: Iterable<string>): string[] {
  const numbers = new Set<string>();
  const pattern = /Merge pull request #(\d+)/iu;
  // Scan each subject for a merge-commit PR reference.
  for (const subj of subjects) {
    const match = pattern.exec(subj);
    if (match) {
      numbers.add(`#${match[1]!}`);
    }
  }
  return [...numbers].sort(compareCodePoint);
}

/**
 * Count conventional commit types in subject lines.
 *
 * Mirrors Python `summarize_conventional_commits`: count each recognized type
 * prefix (`type(`, `type!`, `type:`) else `other`; render `name<9 : value`
 * lines for non-zero counts, or the no-types fallback.
 *
 * @param subjects Newline-separated commit subjects.
 * @returns The conventional-commit summary text.
 */
export function summarizeConventionalCommits(subjects: string): string {
  const counts = new Map<string, number>();
  // Initialize every recognized type plus the "other" bucket to zero.
  for (const key of CONVENTIONAL_TYPES) {
    counts.set(key, 0);
  }
  counts.set("other", 0);

  const typePattern =
    /^(feat|fix|refactor|perf|docs|test|chore|build|ci|style)(\(|!|:)/u;
  // Classify each non-blank subject line by its conventional-commit prefix.
  for (const rawLine of splitLines(subjects)) {
    const line = rawLine.trim();
    if (!line) {
      continue;
    }
    const match = typePattern.exec(line);
    const label = match ? match[1]! : "other";
    counts.set(label, (counts.get(label) ?? 0) + 1);
  }

  // Preserve insertion order (recognized types, then "other"), keeping non-zero.
  const nonZero = [...counts.entries()].filter(([, value]) => value > 0);
  if (nonZero.length === 0) {
    return "(no recognizable conventional commit types)";
  }
  return nonZero
    .map(([name, value]) => `${padName(name)} : ${value}`)
    .join("\n");
}

/**
 * Render structured issue details for the appendix.
 *
 * Mirrors Python `format_issue_details`.
 *
 * @param issue Issue details record.
 * @returns The formatted issue block.
 */
export function formatIssueDetails(issue: IssueDetails): string {
  const commentsText = formatList(issue.comments, "(no comments)");
  const lines = [
    section(`Issue ${issue.number}: ${issue.title}`),
    `State: ${issue.state}`,
    `Author: ${issue.author}`,
    `Labels: ${issue.labels.length > 0 ? issue.labels.join(", ") : "(none)"}`,
    `Assignees: ${issue.assignees.length > 0 ? issue.assignees.join(", ") : "(none)"}`,
    `Created: ${issue.createdAt}`,
    `Updated: ${issue.updatedAt}`,
    "",
    truncate(issue.body, 1200),
    "",
    "Comments:",
    commentsText,
  ];
  if (issue.userStoryContent) {
    lines.push(
      "",
      `User story (${issue.userStoryPath ?? "user-story.md"}):`,
      truncate(issue.userStoryContent, 1200),
    );
  }
  return lines.join("\n");
}

/**
 * Render structured pull request details for the appendix.
 *
 * Mirrors Python `format_pr_details`.
 *
 * @param pr Pull request details record.
 * @returns The formatted PR block.
 */
export function formatPrDetails(pr: PullRequestDetails): string {
  return [
    section(`Pull Request ${pr.number}: ${pr.title}`),
    `State: ${pr.state}`,
    `Author: ${pr.author}`,
    `Base: ${pr.baseRef}`,
    `Head: ${pr.headRef}`,
    `Created: ${pr.createdAt}`,
    `Updated: ${pr.updatedAt}`,
    `Merged: ${pr.mergedAt ?? "(not merged)"}`,
    `Labels: ${pr.labels.length > 0 ? pr.labels.join(", ") : "(none)"}`,
    `Assignees: ${pr.assignees.length > 0 ? pr.assignees.join(", ") : "(none)"}`,
    truncate(pr.body, 1200),
    "",
    "Auto-close issues (from this PR):",
    formatList(pr.closingIssues, "(none)"),
    "",
    "Files (first 15):",
    formatList(pr.filesChanged.slice(0, 15), "(none)"),
  ].join("\n");
}

/**
 * Render the close-candidate section grouped by verification source.
 *
 * Mirrors Python `build_close_candidates_section`.
 *
 * @param params Verified/author-asserted/referenced refs and reason strings.
 * @returns The formatted close-candidates section.
 */
export function buildCloseCandidatesSection(params: {
  verified: string[];
  authorAsserted: string[];
  referenced: string[];
  verifiedReason: string;
  authorReason: string;
}): string {
  const { verified, authorAsserted, referenced, verifiedReason, authorReason } =
    params;
  const allAutoClose = new Set([...verified, ...authorAsserted, ...referenced]);
  const authorAutoClose = [...allAutoClose].sort(compareCodePoint);
  const referencedOnly = referenced
    .filter((ref) => !allAutoClose.has(ref))
    .filter((ref, index, arr) => arr.indexOf(ref) === index)
    .sort(compareCodePoint);

  return [
    section("Close candidates"),
    "Auto-close issues (verified from GitHub PR metadata):",
    formatList(verified, verifiedReason),
    "",
    "Auto-close issues (author asserted):",
    formatList(authorAutoClose, authorReason),
    "",
    "Referenced issues (detected):",
    formatList(referencedOnly, "(none)"),
  ].join("\n");
}

/**
 * Render the approved autoclose section from verified and pending refs.
 *
 * Mirrors Python `build_issues_to_autoclose_section`: verified first, then
 * pending deterministic refs not already verified; when none, the PASS vs
 * non-PASS conservative fallback text.
 *
 * @param params Verified/pending refs and observed readiness signals.
 * @returns The formatted autoclose section.
 */
export function buildIssuesToAutocloseSection(params: {
  verified: string[];
  pendingPrimary: string[];
  readinessSignals: string[];
}): string {
  const { verified, pendingPrimary, readinessSignals } = params;
  // Verified issues first, then pending deterministic issues not yet verified.
  const ordered: string[] = [];
  for (const issue of [...verified, ...pendingPrimary]) {
    if (issue && !ordered.includes(issue)) {
      ordered.push(issue);
    }
  }

  let body: string;
  if (ordered.length > 0) {
    body = formatList(ordered, "(none)");
  } else if (readinessSignals.some((signal) => signal === "PASS")) {
    body =
      "None (no verified closing issues and no deterministic pending issue)";
  } else {
    body = "None (no verified closing issues and readiness not PASS)";
  }

  return [section("Issues to autoclose (verified or pending)"), body].join(
    "\n",
  );
}

/**
 * Extract changed file paths from the 'Changed files' section text.
 *
 * Mirrors Python `extract_changed_paths`: capture lines after a
 * `===== Changed files` header until the next `=====` banner; tab-bearing lines
 * yield the last tab field, otherwise the whole trimmed line, each normalized.
 *
 * @param contextText Full PR context text.
 * @returns The normalized changed file paths.
 */
export function extractChangedPaths(contextText: string): string[] {
  const paths: string[] = [];
  let capture = false;
  // Walk lines, toggling capture at the section header and stopping at the
  // next banner; collect normalized paths from captured lines.
  for (const line of splitLines(contextText)) {
    if (line.startsWith("===== Changed files")) {
      capture = true;
      continue;
    }
    if (capture) {
      if (line.startsWith("=====")) {
        break;
      }
      if (line.trim() && line.includes("\t")) {
        const partsArr = line.split("\t");
        const pathPart = partsArr[partsArr.length - 1]!;
        paths.push(formatDiffPath(pathPart.trim()));
      } else if (line.trim()) {
        paths.push(formatDiffPath(line.trim()));
      }
    }
  }
  return paths;
}

/**
 * Return the final suffix of a path, mirroring Python `Path(name).suffix`.
 *
 * `Path.suffix` is the substring from the last `.` of the final path segment,
 * but only when the dot is not the first character of that segment (so
 * `.gitignore` has no suffix and `archive.tar.gz` yields `.gz`).
 *
 * @param name Path string.
 * @returns The suffix (including the leading dot), or `""`.
 */
function pathSuffix(name: string): string {
  const segment = name.split(/[/\\]/u).pop() ?? "";
  const dotIndex = segment.lastIndexOf(".");
  if (dotIndex <= 0 || dotIndex === segment.length - 1) {
    return "";
  }
  return segment.slice(dotIndex);
}

/** Right-align a count in a width-8 field, mirroring Python `%8d`. */
function padCount(count: number): string {
  return String(count).padStart(8, " ");
}

/** Left-align a name in a width-9 field, mirroring Python `%-9s` (`name:<9`). */
function padName(name: string): string {
  return name.padEnd(9, " ");
}

/** Test whether a string is a non-empty run of ASCII digits (Python isdigit). */
function isDigits(value: string): boolean {
  return value.length > 0 && /^\d+$/u.test(value);
}

/** Compare two strings by Unicode code point (Python `sorted` semantics). */
function compareCodePoint(left: string, right: string): number {
  if (left < right) {
    return -1;
  }
  if (left > right) {
    return 1;
  }
  return 0;
}

/**
 * Split text into lines the way Python `str.splitlines()` does.
 *
 * @param value Text to split.
 * @returns Lines without terminators.
 */
function splitLines(value: string): string[] {
  if (value === "") {
    return [];
  }
  const lines = value.split(/\r\n|\r|\n/u);
  if (
    lines.length > 0 &&
    lines[lines.length - 1] === "" &&
    /(\r\n|\r|\n)$/u.test(value)
  ) {
    lines.pop();
  }
  return lines;
}
